src/curve_intersections.rs

```diff
@@ -0,0 +1,1377 @@
+//! Functions used for clipping parametrized curves and finding intersections between them
+//!
+//! Caution: All functions in this file are provisional and may change in the future!
+
+use crate::common::{solve_cubic, solve_linear, solve_quadratic};
+use crate::{
+    Affine, CubicBez, Line, ParamCurve, ParamCurveArclen, ParamCurveDeriv, ParamCurveExtrema,
+    Point, QuadBez, Rect,
+};
+use arrayvec::ArrayVec;
+use std::ops::Range;
+
+/// A trait for descibing a parameterized curve that can be used in the Bezier clipping algorithm
+pub trait ParamCurveBezierClipping:
+    ParamCurve + ParamCurveDeriv + ParamCurveExtrema + ParamCurveArclen
+{
+    /// Find the time `t` at which the curve has the given x value
+    fn solve_t_for_x(&self, x: f64) -> ArrayVec<f64, 3>;
+    /// Find the time `t` at which the curve has the given x value
+    fn solve_t_for_y(&self, y: f64) -> ArrayVec<f64, 3>;
+    /// Returns the upper and lower convex hull
+    fn convex_hull_from_line(&self, l: &Line) -> (Vec<Point>, Vec<Point>);
+    /// Returns the minimum and maximum distances of the "fat line" enclosing this curve
+    fn fat_line_min_max(&self, baseline: &Line) -> (f64, f64);
+}
+
+// Note that the line is unbounded here!
+fn signed_distance_from_ray_to_point(l: &Line, p: Point) -> f64 {
+    let vec2 = l.p1 - l.p0;
+    let a = -vec2.y;
+    let b = vec2.x;
+    let c = -(a * l.start().x + b * l.start().y);
+
+    let len = (a * a + b * b).sqrt();
```

Comment:
Would be faster to do `.sqrt().recip()` here. I checked .powf(-0.5), which I believe is used elsewhere, and that's surprisingly not optimized, seems to a call into the math library.

I'm surprised because this is well known to be optimizable (both in the famous Quake code and SIMD). But probably that's "fast-math" only, not IEEE compliant. Perhaps we should have a function in common for it.